## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [3]:
import pandas as pd
import numpy as np
import math

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [4]:
# View the column names in the data
print(df.columns)

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')


In [5]:
# Use the describe function to gather some basic statistics
print(df.describe())

          client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.00000

In [6]:
# Use this space to do any additional research
# and familiarize yourself with the data.
# JM Looking for empty data and duplicates
print(df.isnull().sum())
print(df.duplicated().sum())
print(df.info())
#print(df['item_id'].unique())

first          0
last           0
job            0
phone          0
email          0
client_id      0
order_id       0
order_date     0
order_week     0
order_year     0
item_id        0
category       0
subcategory    0
unit_price     0
unit_cost      0
unit_weight    0
qty            0
line_number    0
dtype: int64
0
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 

In [7]:
# What three item categories had the most entries?
category_count = df['category'].value_counts()
category_count.head(3)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [8]:
# For the category with the most entries, which subcategory had the most entries?
# sub_count = df[[df]'consumables'].value_counts()  erronious line correction >
consumables_df = df[df['category'] == 'consumables']
subcat = consumables_df['subcategory'].value_counts()
subcat.head(1)


subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [9]:
# Which five clients had the most entries in the data?
clients = df['client_id'].value_counts()
clients.head(5)


client_id
33615    220
66037    211
46820    209
24741    207
38378    207
Name: count, dtype: int64

In [10]:
# Store the client ids of those top 5 clients in a list.
# creates a list of top 5 clients based on most entries
top_5 = clients.head(5).index.tolist()
print(f"The top 5 clients are: {top_5}")

The top 5 clients are: [33615, 66037, 46820, 24741, 38378]


In [11]:
# How many total units (the qty column) did the client with the most entries order order?
# uses the list reference from the top 5 clients in cell 30 and calls for the first entry
clients = clients.tolist()
most_client_entries = clients[0] # This line caused problems as "clients" is a seriest identified by the helper, and it should be a list GPT recommended converting the series to a list do in the above line
units_ordered = df[df['client_id'] == most_client_entries]['qty'].sum()
most_client_entries
# for readability
print(f"Total units ordered by top client: {most_client_entries}")

Total units ordered by top client: 220


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [12]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['line_subtotal'] = df['unit_price'] * df['qty']
print(df[['unit_price', 'qty', 'line_subtotal']].head(2))

   unit_price  qty  line_subtotal
0     1096.80  105      115164.00
1       24.95   21         523.95


In [13]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
#df['total_weight'] = df['unit_weight'] * df['qty'] #issues with this code, needed to change the order of calulation
#df['shipping_price'] = np.where(df['unit_weight'] > 50, df['unit_weight'] * df['qty'] * 7, df['unit_weight'] *df['qty'] * 10)
df['total_weight'] = df['unit_weight'] * df['qty']
df['shipping_price'] = np.where(df['total_weight'] > 50, df['total_weight'] * 7, df['total_weight'] * 10) # this was a gpt suggestion with the understanding that calculating the total weight first made sure it was available for use in the next line

print(df[['unit_price', 'unit_weight', 'qty', 'total_weight', 'shipping_price']].head(3))


   unit_price  unit_weight  qty  total_weight  shipping_price
0     1096.80         7.50  105        787.50         5512.50
1       24.95         1.49   21         31.29          312.90
2       13.52         1.68   39         65.52          458.64


In [14]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df['line_price'] = (df['line_subtotal'] + df['shipping_price']) * 1.0925
print(df[['line_subtotal', 'shipping_price', 'line_price']].head(3).round(2))

   line_subtotal  shipping_price  line_price
0      115164.00         5512.50   131839.08
1         523.95          312.90      914.26
2         527.28          458.64     1077.12


In [15]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
# there is one less row than the example, 'total_weight' total weight would be used to calculate shipping weight, validating against 'total_price'
df['line_cost'] = (df['unit_cost'] * df['qty']) + df['shipping_price']
df.head(3)


first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   

   line_number  line_subtotal  total_weight  shipping_price     line_price  \
0            1      115164.00        787.50         5512.50  131839.076250   
1            0         523.95         31.29          312.90     914.258625   
2            6         527.28         65.52          458.64    1077.117600   

   line_cost  
0   85597.05  
1     629.79  
2     765.18  

[3 rows x 23 columns]

In [16]:
# Create a column for the profit of each line using line cost and line price
df['profit'] = df['line_price'] - df['line_cost']
df.head(3)




first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   

   line_subtotal  total_weight  shipping_price     line_price  line_cost  \
0      115164.00        787.50         5512.50  131839.076250   85597.05   
1         523.95         31.29          312.90     914.258625     629.79   
2         527.28         65.52          458.64    1077.117600     765.18   

         profit  
0  46242.026250  
1    284.468625  
2    311.937600  

[3 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [17]:
# Check your work using the totals above
#noted descrepancy with inital code indicating rounding issue
#not able to find a rounding rule that satisfied all 3 orders
#tried 
#def round_off(value, decimals=2): #
 #   factor = 10 ** decimals
  #  return math.floor(value * factor) / factor
order_confirmation = [2742071, 2173913, 6128929]
for order_id in order_confirmation:
    line_price = df[df['order_id'] == order_id]['line_price'].sum()
    line_price = round(line_price, 4)
    print(f"Order ID {order_id} has a calculated total price of: ${line_price:,.2f}")

Order ID 2742071 has a calculated total price of: $152,811.90
Order ID 2173913 has a calculated total price of: $162,388.72
Order ID 6128929 has a calculated total price of: $923,441.24


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [18]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
top_5
for client_id in top_5:
    total_spent = df[df['client_id'] == client_id]['line_price'].sum()
    print(f"Client ID {client_id} spent: ${total_spent:,.2f}")




Client ID 33615 spent: $8,377,308.57
Client ID 66037 spent: $10,259,514.80
Client ID 46820 spent: $9,743,794.32
Client ID 24741 spent: $82,268,891.98
Client ID 38378 spent: $12,906,550.88


In [19]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit.
#total cost is not included in the instructions but it is in the example output.  Matching the example output
print(df.columns) 
top_5_df = df[df['client_id'].isin(top_5)]
summary_df = top_5_df.groupby('client_id').agg(
    total_units=('qty', 'sum'),
    total_shipping_price=('shipping_price', 'sum'),
    total_revenue=('line_price', 'sum'),
    total_cost=('line_cost', 'sum'),
    total_profit=('profit', 'sum')
).reset_index()
pd.set_option('display.float_format', '{:.2f}'.format)
summary_df

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number', 'line_subtotal', 'total_weight', 'shipping_price',
       'line_price', 'line_cost', 'profit'],
      dtype='object')


client_id  total_units  total_shipping_price  total_revenue  total_cost  \
0      24741       239862            5126448.37    82268891.98 45688899.71   
1      33615        64313            1828984.89     8377308.57  6175313.91   
2      38378        73667            3429455.40    12906550.88  9634720.98   
3      46820        75768            1601448.84     9743794.32  7007191.64   
4      66037        43018            1395151.85    10259514.80  7004482.98   

   total_profit  
0   36579992.27  
1    2201994.66  
2    3271829.90  
3    2736602.68  
4    3255031.82

In [20]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns. 
money_columns = ['total_shipping_price', 'total_revenue', 'total_cost', 'total_profit']

# Define a function that converts a dollar amount to millions.
def conv_to_millions(value):
    return value / 1_000_000

# Apply the currency_format_millions function to only the money columns. 
summary_df[money_columns] = summary_df[money_columns].apply(conv_to_millions)

# Rename the columns to reflect the change in the money format. 
summary_df.rename(columns={
    'client_id': 'Customer ID',
    'total_shipping_price': 'Shipping (millions)',
    'total_revenue': 'Total Revenue (millions)',
    'total_cost': 'Total Cost (millions)',
    'total_profit': 'Total Profit (millions)',
    'total_units': 'Units'
}, inplace=True)

summary_df

Customer ID   Units  Shipping (millions)  Total Revenue (millions)  \
0        24741  239862                 5.13                     82.27   
1        33615   64313                 1.83                      8.38   
2        38378   73667                 3.43                     12.91   
3        46820   75768                 1.60                      9.74   
4        66037   43018                 1.40                     10.26   

   Total Cost (millions)  Total Profit (millions)  
0                  45.69                    36.58  
1                   6.18                     2.20  
2                   9.63                     3.27  
3                   7.01                     2.74  
4                   7.00                     3.26

In [21]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
sorted_summary_df = summary_df.sort_values(by='Total Profit (millions)', ascending=False).reset_index(drop=True)
sorted_summary_df

Customer ID   Units  Shipping (millions)  Total Revenue (millions)  \
0        24741  239862                 5.13                     82.27   
1        38378   73667                 3.43                     12.91   
2        66037   43018                 1.40                     10.26   
3        46820   75768                 1.60                      9.74   
4        33615   64313                 1.83                      8.38   

   Total Cost (millions)  Total Profit (millions)  
0                  45.69                    36.58  
1                   9.63                     3.27  
2                   7.00                     3.26  
3                   7.01                     2.74  
4                   6.18                     2.20